### DISCUSSION TITLE
ESP32 + ls013b4dn04 2.7" Sharp Memory Display

### DISCUSSION THREAD

**Discussion by Templarian** [Moddable-OpenSource/moddable#921] [Getting started] [answered]
After going through wiring up some buttons and writing a bit of code I ordered the [2.7" Sharp Memory Display](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/displays/wiring-guide-sharp-memory-2.7-spi.md) - [Adafruit Sharp Memory Display](https://www.adafruit.com/product/4694). I have a few ESP32 boards, but for this I'm using the `esp32/esp32_thing` as it is generally in stock and well supported.

There are examples for the `esp` / `gecko` devices, so I'm thinking maybe it might not work yet. Should this work out of the box or am I doing something wrong below:

manifest.json:

```json
{
	"include" : [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODULES)/drivers/ls013b4dn04/manifest.json",
		"$(MODULES)/base/instrumentation/manifest.json"
	],
	"config": {
		"screen": "ls013b4dn04",
		"touch": "",
		"format": "Gray256"
	},
	"resources": {
		"*": [
			"./moddable-white"
		]
	},
	"defines": {
		"ls013b4dn04": {
			"width": 400,
			"height": 240,
			"hz": 3500000,
			"cs_pin": 4,
			"spi_port": 23
		}
	},
	"modules": {
		"commodetto/Poco": "$(COMMODETTO)/commodettoPoco",
		"commodetto/*": "$(COMMODETTO)/commodettoPocoBlit",
		"commodetto/parseBMF": "$(COMMODETTO)/commodettoParseBMF",
		"commodetto/parseBMP": "$(COMMODETTO)/commodettoParseBMP",
		"commodetto/Bitmap": "$(COMMODETTO)/commodettoBitmap",
		"Resource": "$(MODDABLE)/modules/files/resource/Resource",
		"*": "./main"
	},
	"preload": [
		"commodetto/*",
		"Resource",
		"instrumentation",
		"pins",
		"spi",
		"main"
	]
}
```

main.js:

```js
import LS013B4DN04 from "ls013b4dn04";
import Poco from "commodetto/Poco";
import Resource from "Resource";
import parseBMP from "commodetto/parseBMP";
import Timer from "timer";

debugger;

const TIMER_MS = 5000;

export default function() {
    // Took this example from the moddable example for ls013b4dn04 (removing sleep)
    const logo = parseBMP(new Resource("moddable-white.bmp"));

    const width = 400, height = 240;
    let render = new Poco(new LS013B4DN04({ width: width, height: height }));

    let black = render.makeColor(0, 0, 0);
    let white = render.makeColor(255, 255, 255);

    let index = 0;

     let timer = Timer.repeat(() => {
        index ^= 1;
        render.begin(0, 0, width, height);
            render.fillRectangle(index ? white : black, 0, 0, width, height);
            render.drawGray(logo, index ? black : white, 0, 30);
        render.end();
    }, TIMER_MS);
}
```

Run:

```
mcconfig -d -m -p esp32/esp32_thing
```

Error in the terminal:

```
Executing action: all (aliases: build)
Running ninja in directory c:\users\admin\projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsproj-esp32\build
Executing "ninja all"...
[2/7] Linking CXX executable xs_esp32.elf
FAILED: xs_esp32.elf
```

<details>
  <summary>Click to view the full output...</summary>

```
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
Reconfiguring ESP-IDF...
Executing action: reconfigure
Running cmake in directory c:\users\admin\projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsproj-esp32\build
Executing "cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -DSDKCONFIG_DEFAULTS=C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/sdkconfig.mc -DSDKCONFIG_HEADER=C:\Users\admin\Projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsProj-esp32\build\config\sdkconfig.h -DCMAKE_MESSAGE_LOG_LEVEL=ERROR -DDEBUGGER_SPEED=460800 -DIDF_TARGET=esp32 -DESP32_SUBCLASS=esp32 -DSDKCONFIG_DEFAULTS=C:\Users\admin\Projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsProj-esp32\sdkconfig.mc -DCCACHE_ENABLE=1 c:\users\admin\projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsproj-esp32"...
Loading defaults file C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/sdkconfig.mc...
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/build
        1 file(s) copied.
# xsl modules
### 295 instances, 591 keys, 62 colors, 0 holes
# cc mc.xs.o (slots in flash)
# mcrez resources
Total resource size: 484 bytes
# cc mc.resources.o (slots in flash)
# ld xs_esp32.bin
        1 file(s) copied.
        1 file(s) copied.
Executing action: all (aliases: build)
Running ninja in directory c:\users\admin\projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsproj-esp32\build
Executing "ninja all"...
[2/7] Linking CXX executable xs_esp32.elf
FAILED: xs_esp32.elf
cmd.exe /C "cd . && C:\Espressif\tools\xtensa-esp32-elf\esp-2021r2-patch3-8.4.0\xtensa-esp32-elf\bin\xtensa-esp32-elf-g++.exe -mlongcalls -Wno-frame-address  @CMakeFiles\xs_esp32.elf.rsp -o xs_esp32.elf  && cd ."
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `pocoGetBitmapPixels':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:979: undefined reference to `CFENew'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_build':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:87: undefined reference to `CFENew'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_close':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:97: undefined reference to `CFELockCache'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_begin':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:147: undefined reference to `CFELockCache'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_end':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:298: undefined reference to `CFELockCache'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_drawFrame':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:694: undefined reference to `CFESetFontData'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:694: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_getTextWidth':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:699: undefined reference to `CFESetFontData'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:712: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(commodettoPoco.c.o): in function `xs_poco_drawText':
C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:740: undefined reference to `CFESetFontData'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:741: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:758: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:763: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:764: undefined reference to `CFEGetGlyphFromUnicode'
c:/espressif/tools/xtensa-esp32-elf/esp-2021r2-patch3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: C:\Users\admin\Projects\moddable\modules\commodetto/commodettoPoco.c:888: undefined reference to `CFEGetGlyphFromUnicode'
collect2.exe: error: ld returned 1 exit status
[3/7] Performing build step for 'bootloader'
[0/1] Re-running CMake...
-- Building ESP-IDF components for target esp32
-- Project sdkconfig file C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/sdkconfig
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/soc/esp32/ld/esp32.peripherals.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.api.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.libgcc.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom/esp32/ld/esp32.rom.newlib-funcs.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main/ld/esp32/bootloader.ld
-- Adding linker script C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main/ld/esp32/bootloader.rom.ld
-- Components: bootloader bootloader_support efuse esp32 esp_common esp_hw_support esp_rom esp_system esptool_py freertos hal log main micro-ecc newlib partition_table soc spi_flash xtensa
-- Component paths: C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader_support C:/Espressif/frameworks/esp-idf-v4.4.1/components/efuse C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp32 C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_common C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_hw_support C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_rom C:/Espressif/frameworks/esp-idf-v4.4.1/components/esp_system C:/Espressif/frameworks/esp-idf-v4.4.1/components/esptool_py C:/Espressif/frameworks/esp-idf-v4.4.1/components/freertos C:/Espressif/frameworks/esp-idf-v4.4.1/components/hal C:/Espressif/frameworks/esp-idf-v4.4.1/components/log C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/main C:/Espressif/frameworks/esp-idf-v4.4.1/components/bootloader/subproject/components/micro-ecc C:/Espressif/frameworks/esp-idf-v4.4.1/components/newlib C:/Espressif/frameworks/esp-idf-v4.4.1/components/partition_table C:/Espressif/frameworks/esp-idf-v4.4.1/components/soc C:/Espressif/frameworks/esp-idf-v4.4.1/components/spi_flash C:/Espressif/frameworks/esp-idf-v4.4.1/components/xtensa
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/build/bootloader
[1/1] cmd.exe /C "cd /D C:\Users\admin\Projects\moddable\build\tmp\esp32\esp32_thing\debug\ls013b4dn04\xsProj-esp32\build\bootloader\esp-idf\esptool_py && C:\Espressif\python_env\idf4.4_py3.8_env\Scripts\python.exe C:/Espressif/frameworks/esp-idf-v4.4.1/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x1000 C:/Users/admin/Projects/moddable/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/xsProj-esp32/build/bootloader/bootloader.bin"
Bootloader binary size 0x4450 bytes. 0x2bb0 bytes (39%) free.
ninja: build stopped: subcommand failed.
ninja failed with exit code 1
```

</details>

**Comment by phoddie** [comment, marked as answer]
The Commodetto Font Engine (CFE) isn't being included, which is causing a link error The easiest way to resolve that is to include `manifest_commodetto.json`. Here's a revisited manifest that builds for ESP8266 (and should work for ESP32 as well).

```json
{
	"include" : [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_commodetto.json",
		"$(MODULES)/drivers/ls013b4dn04/manifest.json"
	],
	"config": {
		"screen": "ls013b4dn04",
		"touch": "",
		"format": "Gray256"
	},
	"resources": {
		"*": [
			"./moddable-white"
		]
	},
	"defines": {
		"ls013b4dn04": {
			"width": 96,
			"height": 96,
			"hz": 3500000
		}
	},
	"modules": {
		"Resource": "$(MODDABLE)/modules/files/resource/Resource",
		"*": "./main"
	},
	"preload": [
		"Resource",
		"main"
	],
	"strip":"*"
}
```

**Reply by Templarian** [reply]
That builds and deploys the code. **But** nothing displays on the screen... and putting `debugger;` line doesn't hit any breakpoints. Weirdly the `main` tab keeps flashing, but can't click on it in the debugger. I switch over to the `HSPI` pins in an updated manifest below.

![image](https://user-images.githubusercontent.com/338885/177404659-90d86c9c-9893-4a97-ac1a-153f9108fe90.png)

I have all 5 pins connected. Looking at the [v1.0 ESP32 Thing diagram](https://cdn.sparkfun.com/assets/learn_tutorials/5/0/7/ESP32ThingV1a.pdf).

| Pin | Description | Name on Display |
|----|--------|--------|
| 3.3 | Power | VIN (this board has 3.3 -> 5) |
| GND | Ground | GND |
| 15 | `ls013b4dn04.cs_pin`. "CS is active high!" | CS |
| 13 | `ls013b4dn04.spi_port`  | MOSI |
| 14 | `HSPI_CLK` Listed in [diagram](https://cdn.sparkfun.com/assets/learn_tutorials/5/0/7/ESP32ThingV1a.pdf). | SCLK |

Note: Even if I simplify the example below I cannot get it to break on the debugger lines and the "main" tab keeps flashing in the debugger window. I added the blink onboard LED lines and it's not blinking (this works in the hello world app).

```javascript
import LS013B4DN04 from "ls013b4dn04";
import Poco from "commodetto/Poco";
import Resource from "Resource";
import parseBMP from "commodetto/parseBMP";
import Timer from "timer";
import Digital from "pins/digital";

const TIMER_MS = 5000;
const ONBOARD_LED_PIN = 5;

export default function() {

	// Let me know it's running
	let count = 0;
	Timer.repeat(() => {
		++count;
		Digital.write(ONBOARD_LED_PIN, count & 1);
	}, 500);

	trace('App running?');
	debugger;

	const logo = parseBMP(new Resource("moddable-white.bmp"));

	const width = 400, height = 240;
	let render = new Poco(new LS013B4DN04({ width: width, height: height }));

	let black = render.makeColor(0, 0, 0);
	let white = render.makeColor(255, 255, 255);

	let index = 0;

	let timer = Timer.repeat(() => {
		index ^= 1;
		render.begin(0, 0, width, height);
			render.fillRectangle(index ? white : black, 0, 0, width, height);
			render.drawGray(logo, index ? black : white, 0, 30);
		render.end();
	}, TIMER_MS);

}
```

```json
{
	"include" : [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_commodetto.json",
		"$(MODULES)/pins/digital/manifest.json",
		"$(MODULES)/drivers/ls013b4dn04/manifest.json"
	],
	"config": {
		"screen": "ls013b4dn04",
		"touch": "",
		"format": "Gray256"
	},
	"resources": {
		"*": [
			"./moddable-white"
		]
	},
	"defines": {
		"ls013b4dn04": {
			"width": 400,
			"height": 240,
			"hz": 3500000,
			"cs_pin": 15,
			"spi_port": 13
		}
	},
	"modules": {
		"Resource": "$(MODDABLE)/modules/files/resource/Resource",
		"*": "./main"
	},
	"preload": [
		"Resource",
		"main"
	],
	"strip":"*"
}
```

Feel like I'm close to having this solved but tried a lot of ideas, but I can't get it to hit `debugger` points and the LED isn't blinking so I think it's not running.

**Reply by phoddie** [reply]
Congratulations on the progress. Thank you for sharing the update.

> the "main" tab keeps flashing in the debugger window

That suggests that the device is constantly rebooting (it connects to the debugger, runs for a bit, then restarts). That is likely due to something failing in the setup code before your main.js script runs. You can confirm that be setting Break On Start in the xsbug preferences.

My guess is that it is failing to initialize the screen in the [Piu setup code](https://github.com/Moddable-OpenSource/moddable/blob/886694700ec392b3c6f92cfa92cf47d8ff0b178f/build/devices/esp32/setup/piu.js#L198). That's because the `spi_port` in the config is set to `13`, which is an invalid value on ESP32. The usual value for the display SPI port on ESP32 is `"HSPI_HOST"` as set-up on [Moddable Two for the ILI9341](https://github.com/Moddable-OpenSource/moddable/blob/886694700ec392b3c6f92cfa92cf47d8ff0b178f/build/devices/esp32/targets/moddable_two/manifest.json#L32), for example, or `"VSPI_HOST"` as used on some M5Stack devices.

**Reply by Templarian** [reply]
Awesome! Code is all running now and breakpoints hit. Thanks. Now trying to figure out why the display is showing nothing still, but at least the code is executing.

I also have the smaller [1.3 Sharp Memory display](https://www.adafruit.com/product/3502) I can give a try incase the display is dead. It looks like this should be working. Tried some other drawing code and nothing shows up yet.

**Reply by Templarian** [reply]
Marking your first reply as the answer, but will keep updating here until I get something rendering.

/edit it's not working with the smaller 168x144. So will keep trying ideas.

**Reply by Templarian** [reply]
@phoddie First let me say how much I appreciate your help and I got it to work! But...

Kind of hacked it by doing a `Digital.write(15, 1);`. Since CS is active high it's deactivating the screen when it should be activating it as far as I can tell. But I'm not really sure I fully understand this enough to fix it correctly.

- https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/ls013b4dn04/modLs013b4dn04.c#L59
  - I think it has to do with this line.

I'll make a PR to add wiring for the ESP32 / adafruit breakouts once I have this working fully.

**Reply by phoddie** [reply]
Very cool that it works. I know it can be frustrating working with unfamiliar software to bring up new hardware.

CS is active high in the implementation:

```c
#define SCREEN_CS_DEACTIVE	modGPIOWrite(&ls->cs, 0)
#define SCREEN_CS_ACTIVE	modGPIOWrite(&ls->cs, 1)
```

All the ls013b4dn04 display modules we've seen are CS high. The line you reference sets the pin to inactive at initialization, which seems the correct state until there's something to write. Where did you add the `Digital.write(15, 1)`?

> I'll make a PR to add wiring for the ESP32 / adafruit breakouts once I have this working fully.

That would be great, thank you. FYI – Moddable needs a [CLA](https://github.com/Moddable-OpenSource/moddable/tree/public/licenses#contributor-license-agreement) (Contributor License Agreement) to merge the PR.

**Reply by Templarian** [reply]
I added a button to test what happens if I clicked it and afterwards the screen starts to render. Still playing around to get the larger 2.7" screen to work (but I may have broken this one).

```
const button = new Digital(ONBOARD_BUTTON_PIN, Digital.InputPullUp);
    let previous = 1;
    Timer.repeat(() => {
        const current = button.read();
        // trace(`${current} !== ${previous}\n`);
        if (current !== previous) {
            if (current) {
                trace("force CS button to active\n");
                Digital.write(15, 1);
            }
            previous = current;
        }
    }, 100);
```

**Reply by phoddie** [reply]
Interesting. I wonder if somehow the CS pin definition ends up being wrong. It is set in your manifest. It is also set (to a different number) in the [driver manifest](https://github.com/Moddable-OpenSource/moddable/blob/acb12341204a754378608e9cc0d0a937903eacfe/modules/drivers/lpm013m126a/manifest.json#L41-L45). Your script's definition should override the driver. Maybe check your build results to see that it is set correct. The relevant file is `$MODDABLE/build/tmp/esp32/esp32_thing/debug/ls013b4dn04/mc.defines.h` (the middle part depends on the platform you are building for). We expect to see `15` for `MODDEF_LS013B4DN04_CS_PIN` but perhaps it is `4` from the driver's manifest?

**Reply by Templarian** [reply]
Pasted below it looks fine. It is including `MODDEF_ILI9341_*` stuff, but that might be normal.

(offtopic: <strike>the 2.7" display I have is a little broken due to a bad ribbon cable that just won't sit correctly</strike>, edit: fixed with superglue)

- Seeing the same issue with the 2.7" and 1.3". The `Digital.write(15, 1);` and then it starts rendering, but I don't know why.

```
/* WARNING: This file is automatically generated. Do not edit. */

#define MODDEF_ILI9341_WIDTH (240)
#define MODDEF_ILI9341_HEIGHT (320)
#define MODDEF_ILI9341_HZ (40000000)
#define MODDEF_ILI9341_FLIPX (false)
#define MODDEF_ILI9341_FLIPY (false)
#define MODDEF_ILI9341_CS_PIN (15)
#define MODDEF_ILI9341_DC_PIN (2)
#define MODDEF_ILI9341_SPI_PORT HSPI_HOST
#define MODDEF_I2C_SDA_PIN (21)
#define MODDEF_I2C_SCL_PIN (22)
#define MODDEF_SPI_MISO_PIN (12)
#define MODDEF_SPI_MOSI_PIN (13)
#define MODDEF_SPI_SCK_PIN (14)
#define MODDEF_LS013B4DN04_WIDTH (400)
#define MODDEF_LS013B4DN04_HEIGHT (240)
#define MODDEF_LS013B4DN04_CS_PIN (15)
#define MODDEF_LS013B4DN04_DC_PIN (2)
#define MODDEF_LS013B4DN04_SPI_PORT HSPI_HOST
#define PIU_DASH_SIGNATURE "tech-moddable-ls013b4dn04"
#define PIU_DOT_SIGNATURE "tech.moddable.ls013b4dn04"
#define PIU_SLASH_SIGNATURE "/tech/moddable/ls013b4dn04"
```

https://user-images.githubusercontent.com/338885/177465054-6d6fb7aa-b985-4826-8ee7-6f4e5af3b6ef.mp4

**Reply by phoddie** [reply]
Thanks for the video! The big screen is great. I really like how quickly it refreshes.

The dump of `mc.defines.h` shows that `MODDEF_LS013B4DN04_CS_PIN` is correctly set to 15. So, my guess was wrong. But, it also shows `MODDEF_ILI9341_CS_PIN` set to 15. I think you may be running two display drivers – the ILI9341 from the esp32/esp32thing manifest and the ls013b4dn04 from your manifest. The ILI9341 uses CS with active low, which might explain why it is being driven low unexpectedly.

You can override the use of ILI9341 by in the `config` section of your project's manifest by setting the screen driver:

```json
	"config": {
		"screen": "ls013b4dn04"
	}
```

Since you are including `manifest_commodetto.json`,  `screen` global will automatically be initialized with an instance of the ls013b4dn04 driver. So, instead of calling `new LS013B4DN04` in your main, you would just use `screen` instead.  Alternatively, you could set `"screen"` to `""` in the manifest config section. Then your main should run unchanged, but the `screen` global won't be set-up.

**Reply by Templarian** [reply]
It took a bit to test, long day then started playing around with ideas to make it work.

- I have `"screen": "ls013b4dn04"` set already.
- Looks like it's merging the target `build\devices\esp32\targets\esp32_thing\manifest.json` file. 
- I removed the `screen = new LS013B4DN04()` line and it's picking up the global. 😄 
- If I delete the target files `config`/`include` properties they don't show in the `mc.defines.h`, so I think it's merging them.
    - the display still doesn't work without putting `Digital.write(15, 1);` before the render logic.

By the way I'm running `mcconfig -d -m -p esp32/esp32_thing`

```
{
	"include" : [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_commodetto.json",
		"$(MODULES)/drivers/ls013b4dn04/manifest.json"
	],
	"config": {
		"screen": "ls013b4dn04",
		"touch": "",
		"format": "Gray256"
	},
	"resources": {
		"*": [
			"./moddable-white"
		]
	},
	"defines": {
		"ls013b4dn04": {
			"width": 400,
			"height": 240,
			"cs_pin": 15,
			"spi_port": "HSPI_HOST"
		}
	},
	"modules": {
		"Resource": "$(MODDABLE)/modules/files/resource/Resource",
		"*": "./main"
	},
	"preload": [
		"Resource",
		"main"
	],
	"strip":"*"
}
```

**Reply by phoddie** [reply]
Thanks for the additional info. Good that we can rule interference from the ILI9341 driver out.

I don't have a display connected to test with, but I ran a few tests on the ESP32 today to see the behavior of the CS pin (GPIO 15). It behaves as expected, of course. ;) I tried to puzzle out what your change to set GPIO 15 high outside the driver. The only effect that I can see is that it causes GPIO 15 to be set high earlier than it would otherwise.

You are running on an ESP32, which is relatively fast compared to the ESP8266 and Gecko hardware we have used with the Sharp Memory Displays. Checking the data sheet for the ls013b4dn04, there is a relatively long time required for CS to settle.

<img width="626" alt="image" src="https://user-images.githubusercontent.com/6856458/177849346-994a7118-f40e-4079-b93c-8d47cff63ffc.png">

Maybe on the faster ESP32, there isn't a long enough elapsed time between setting CS high and beginning to transmit pixel data. Your change effectively extends that. A quick hack can test the theory: in the driver, wait a little after setting CS high. This change waits much longer than should be necessary:

```c
void ls013b4dn04ChipSelect(uint8_t active, modSPIConfiguration config)
{
	ls013b4dn04 ls = (ls013b4dn04)(((char *)config) - offsetof(ls013b4dn04Record, spiConfig));

	if (active) {
		SCREEN_CS_ACTIVE;
		modDelayMilliseconds(2);
	}
	else
		SCREEN_CS_DEACTIVE;
}
```

**Reply by Templarian** [reply]
@phoddie Does not fix the issue, but does create a sort of 1 second render effect from top to bottom filling in rows. This still requires me to click the button once to send the `Digital.write(15, 1);` (ignore the display showing something initially that is from the last run, neat feature of the memory display 😃)

https://user-images.githubusercontent.com/338885/177862215-be98a069-d728-4b5b-92ad-51c33cd4a30e.mp4

.

**Comment by phoddie** [comment]
Too bad. The super slo-mo is nice though.

I'm out of ideas. You should at least be able to move your workaround out of script and into the driver. At the end of the native constructor, `xs_LS013B4DN04`, just add this:

```c
SCREEN_CS_ACTIVE;
```

The call to `ls_clear()` just before that leaves CS set to 0 (inactive).

**Reply by Templarian** [reply]
Weird enough but that also didn't work either. NBD, you got everything working enough for my project! Very excited to start coding more on this (connecting a i2c fuel gauge next while designing a UI for this mono display).

One suggestion though I think it would make sense to remove the default screen and the include from the `esp32_thing` target. I assume this gets stripped out eventually, but it looks like any defines do get included as we saw.

- `build\devices\esp32\targets\esp32_thing\manifest.json`

I'll sign the CLA and make a PR to update the docs with more details for ESP32 / Wiring for these 2 adafruit breakouts.

**Reply by phoddie** [reply]
> Weird enough but that also didn't work either

Very strange. There's some mystery here. When you figure it out, please let me know. ;)

> Very excited to start coding more on this (connecting a i2c fuel gauge)

There are many ways to implement and integrate sensors, like an I²C fuel gauge. The Moddable SDK is moving in the direction of using the [Ecma-419](https://419.ecma-international.org) for that. It defines IO for both pure [I²C](https://419.ecma-international.org/#-io-classes-ic) and [SMBus](https://419.ecma-international.org/#-io-classes-system-management-bus-smbus). The [Sensor Class Pattern](https://419.ecma-international.org/#-sensor-classes) specifies a common API for all sensors with extensibility. We have a [collection](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/peripherals) of compatible sensor drivers with more on the way. Anyway... do whatever works for you, but we've found the Ecma-419 stuff work well and packages the implementation for easy re-use. All the Ecma-419 foundation bits are already in the Moddable SDK.

> One suggestion though I think it would make sense to remove the default screen and the include from the esp32_thing target

That makes sense. The ESP32-Thing doesn't have a built-in display. I think the reason it is there is from our testing when bringing up the port. Do you want to make that chance?

> I'll sign the CLA

Thank you. We received that.

**Reply by Templarian** [reply]
Will mess around more debugging the `c` side later. I do eventually want to remove the workaround.

I'll submit a PR to update the target for the `esp32_thing` and any docs/examples this weekend.

Thanks for the additional info for the I²C stuff and Ecma-419. I'm sure as I get into this more it will be more clear (little bit of a learning curve).
